src/tools/quemap/lightgrid.c

```diff
@@ -433,6 +433,176 @@ void IndirectLightgrid(int32_t luxel_num) {
+            i1 = 0; j1 = 1; k1 = 0; i2 = 1; j2 = 1; k2 = 0; // Y X Z order
+        }
+    }
+
+    // A step of (1,0,0) in (i,j,k) means a step of (1-c,-c,-c) in (x,y,z),
+    // a step of (0,1,0) in (i,j,k) means a step of (-c,1-c,-c) in (x,y,z), and
+    // a step of (0,0,1) in (i,j,k) means a step of (-c,-c,1-c) in (x,y,z), where
+    // c = 1/6.
+    float x1 = x0 - i1 + G3; // Offsets for second corner in (x,y,z) coords
+    float y1 = y0 - j1 + G3;
+    float z1 = z0 - k1 + G3;
+    float x2 = x0 - i2 + 2.0f * G3; // Offsets for third corner in (x,y,z) coords
+    float y2 = y0 - j2 + 2.0f * G3;
+    float z2 = z0 - k2 + 2.0f * G3;
+    float x3 = x0 - 1.0f + 3.0f * G3; // Offsets for last corner in (x,y,z) coords
+    float y3 = y0 - 1.0f + 3.0f * G3;
+    float z3 = z0 - 1.0f + 3.0f * G3;
+
+    // Work out the hashed gradient indices of the four simplex corners
+    int32_t gi0 = SimplexHash(i + SimplexHash(j + SimplexHash(k)));
+    int32_t gi1 = SimplexHash(i + i1 + SimplexHash(j + j1 + SimplexHash(k + k1)));
+    int32_t gi2 = SimplexHash(i + i2 + SimplexHash(j + j2 + SimplexHash(k + k2)));
+    int32_t gi3 = SimplexHash(i + 1 + SimplexHash(j + 1 + SimplexHash(k + 1)));
+
+    // Calculate the contribution from the four corners
+    float t0 = 0.6f - x0*x0 - y0*y0 - z0*z0;
+    if (t0 < 0) {
+        n0 = 0.0;
+    } else {
+        t0 *= t0;
+        n0 = t0 * t0 * SimplexGrad(gi0, x0, y0, z0);
+    }
+    float t1 = 0.6f - x1*x1 - y1*y1 - z1*z1;
+    if (t1 < 0) {
+        n1 = 0.0;
+    } else {
+        t1 *= t1;
+        n1 = t1 * t1 * SimplexGrad(gi1, x1, y1, z1);
+    }
+    float t2 = 0.6f - x2*x2 - y2*y2 - z2*z2;
+    if (t2 < 0) {
+        n2 = 0.0;
+    } else {
+        t2 *= t2;
+        n2 = t2 * t2 * SimplexGrad(gi2, x2, y2, z2);
+    }
+    float t3 = 0.6f - x3*x3 - y3*y3 - z3*z3;
+    if (t3 < 0) {
+        n3 = 0.0;
+    } else {
+        t3 *= t3;
+        n3 = t3 * t3 * SimplexGrad(gi3, x3, y3, z3);
+    }
+    // Add contributions from each corner to get the final noise value.
+    // The result is scaled to stay just inside [-1,1]
+    return 32.0f*(n0 + n1 + n2 + n3);
+}
+
+/**
+ * Fractal/Fractional Brownian Motion (fBm) summation of 3D Perlin Simplex noise
+ *
+ * @param[in] octaves   number of fraction of noise to sum
+ * @param[in] x         x float coordinate
+ * @param[in] y         y float coordinate
+ * @param[in] z         z float coordinate
+ *
+ * @return Noise value in the range[-1; 1], value of 0 on all integer coordinates.
+ */
+float SimplexNoiseFBM(size_t octaves, float frequency, float amplitude, float lacunarity, float persistence, float x, float y, float z) {
+    float output = 0.f;
+    float denom  = 0.f;
+
+    for (size_t i = 0; i < octaves; i++) {
+        output += (amplitude * SimplexNoise(x * frequency, y * frequency, z * frequency));
+        denom += amplitude;
+
+        frequency *= lacunarity;
+        amplitude *= persistence;
+    }
+
+    return (output / denom);
+}
+
```

Comment:
Could we move all of this to its own file, like what we did for the ambient occlusion shapes? Just to keep the lightgrid code a bit more focused / readable.

Reply:
Done. Simplex.h/simplex.c in Quemap. No other lib will use it I don't think so I just isolated it to there.